runtime/compiler/z/codegen/J9S390PrivateLinkage.cpp

```diff
@@ -3051,6 +3051,97 @@ void TR::J9S390JNILinkage::acquireVMAccessMask(TR::Node * callNode, TR::Register
    }
 
 
+#ifdef J9VM_INTERP_ATOMIC_FREE_JNI
+void
+TR::J9S390JNILinkage::releaseVMAccessMaskAtomicFree(TR::Node * callNode,
+                                                    TR::Register * methodMetaDataVirtualRegister,
+                                                    TR::Register * tempReg1)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)fe();
+   TR::CodeGenerator* cg = self()->cg();
+   TR::Compilation* comp = self()->comp();
+
+   // Store 1 into vmthread->inNative
+   generateRIInstruction(cg, TR::InstOpCode::getLoadHalfWordImmOpCode(), callNode, tempReg1, 1);
+   generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), callNode, tempReg1,
+                         generateS390MemoryReference(methodMetaDataVirtualRegister, offsetof(J9VMThread, inNative), cg));
+
+#if !defined(J9VM_INTERP_ATOMIC_FREE_JNI_USES_FLUSH)
+   // The asymmetric algorithm does not need this serialization.
+   TR::InstOpCode::S390BranchCondition cond = TR::InstOpCode::COND_MASK14;
+   if (!cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_z196))
+      cond = TR::InstOpCode::COND_MASK15;
+
+   TR::RealRegister *gpr0 = cg->machine()->getS390RealRegister(TR::RealRegister::GPR0);
+   new (cg->trHeapMemory()) TR::S390RegInstruction(TR::InstOpCode::BCR, callNode, cond, gpr0, cg);
```

Comment:
Can we write a new `generate*` helper for this one?

Reply:
Fixed. I should've searched a bit more carefully. We already have a helper for this.